src/app/layout/qgslayoutlegendwidget.cpp

```diff
@@ -900,6 +903,44 @@ void QgsLayoutLegendWidget::mExpressionFilterButton_toggled( bool checked )
   mLegend->endCommand();
 }
 
+void QgsLayoutLegendWidget::mLayerExpressionButton_clicked()
+{
+
+  if ( !mLegend )
+  {
+    return;
+  }
+
+  QModelIndex currentIndex = mItemTreeView->currentIndex();
+  if ( !currentIndex.isValid() )
+    return;
+
+  QgsLayerTreeNode *currentNode = mItemTreeView->currentNode();
+  if ( !QgsLayerTree::isLayer( currentNode ) )
+    return;
+
+  QgsLayerTreeLayer *layer = qobject_cast<QgsLayerTreeLayer *>( currentNode );
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( layer );
+
+  if ( !vl )
```

Comment:
code fails here, unsure about what might cause the issues and if there is a better way to get the vectorlayer or if the condition itself is bad.

Reply:
@m-kuhn Since I need to obtain a vector layer for the expression builder I am using the double cast, would it be better to remove line 922 and try directly line 923?

Could the double object_cast break something and cause line 925 to evaluate as True?

Reply:
QgsVectorLayer does not inherit from QgsLayerTreeLayer.
Use `QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( layer->layer() );` instead (and better rename layer to layerNode on line 922 for clarity)

Reply:
@m-kuhn Thanks! I forgot about that inheritance and how it worked with qobject_cast.